erigon-lib/compress/silkworm_seg_fuzz_test.go

```diff
@@ -0,0 +1,223 @@
+			if appendErr != nil {
+				return appendErr
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		words.CloseAndRemove()
+		return err
+	}
+
+	words.Close()
+	return nil
+}
+
+func SegZipSilkworm(path string, cmdPath string) error {
+	cmd := exec.Command(cmdPath, "seg_zip", path)
+	return cmd.Run()
+}
+
+type RandPattern struct {
+	pattern []byte
+}
+
+func NewRandPattern(r *rand.Rand, patternLen int) RandPattern {
+	pattern := make([]byte, patternLen)
+	r.Read(pattern)
+	return RandPattern{pattern}
+}
+
+func (p RandPattern) CopyTo(word []byte, offset int) {
+	copy(word[offset:cmp.Min(offset+len(p.pattern), len(word))], p.pattern)
+}
+
+func generatePatterns(r *rand.Rand) []RandPattern {
+	const minPatternLen = 64
+	const maxPatternLen = 256
+	const patternsCount = 16
+
+	patterns := make([]RandPattern, 0, patternsCount)
+	for i := 0; i < patternsCount; i++ {
+		patternLen := r.Intn(maxPatternLen+1-minPatternLen) + minPatternLen
+		pattern := NewRandPattern(r, patternLen)
+		patterns = append(patterns, pattern)
+	}
+	return patterns
+}
+
+func generateRawWordsFile(path string, seed int64) (*DecompressedFile, error) {
+	const maxTotalSize = int(512 * datasize.KB)
+	const maxWordLen = int(1 * datasize.KB)
+	const maxWordPatterns = 3
+
+	words, err := NewUncompressedFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	r := rand.New(rand.NewSource(seed))
+	totalSizeLimit := maxTotalSize // r.Intn(maxTotalSize + 1)
+	var wordLen int
+
+	patterns := generatePatterns(r)
+
+	for totalSize := 0; totalSize < totalSizeLimit; totalSize += wordLen + 1 {
+		// generate a random word
+		wordLen = r.Intn(maxWordLen + 1)
+		word := make([]byte, wordLen)
+		r.Read(word)
+
+		// fill it with some patterns at random offsets
+		if wordLen > 0 {
+			patternsCount := r.Intn(maxWordPatterns + 1)
+			for i := 0; i < patternsCount; i++ {
+				pattern := patterns[r.Intn(len(patterns))]
+				offset := r.Intn(wordLen)
+				pattern.CopyTo(word, offset)
+			}
+		}
+
+		isCompressed := r.Intn(100) > 0
+		if isCompressed {
+			err = words.Append(word)
+		} else {
+			err = words.AppendUncompressed(word)
+		}
+		if err != nil {
+			words.CloseAndRemove()
+			return nil, err
+		}
+	}
+
+	err = words.Flush()
+	if err != nil {
+		words.CloseAndRemove()
+		return nil, err
+	}
+
+	return words, nil
+}
+
+func copyFiles(sourceFilePaths []string, targetDirPath string) {
+	if len(targetDirPath) == 0 {
+		return
+	}
+	for _, path := range sourceFilePaths {
+		_ = exec.Command("cp", path, targetDirPath).Run()
+	}
+}
+
+func FuzzSilkwormCompress(f *testing.F) {
+	ctx := context.Background()
+	logger := log.New()
+
+	cmdPath, _ := os.LookupEnv("SILKWORM_SNAPSHOTS_CMD")
+	require.NotEmpty(f, cmdPath, "Build silkworm snapshots command and set SILKWORM_SNAPSHOTS_CMD environment variable to the executable path, e.g. $HOME/silkworm/build/cmd/dev/snapshots")
+	investigationDir, _ := os.LookupEnv("INVESTIGATION_DIR")
+
+	f.Add(int64(0))
+	f.Fuzz(func(t *testing.T, seed int64) {
```

Comment:
i think it's a bit anti-pattern: to use input of fuzz test as random seed. because fuzzing framework trying to mutate input to increase code coverage - but "any meaningful seed mutation" will change code coverage randomly(or likely will not change at all - because 1 random data-set will look very similar to another random data-set). 

Better give more control for fuzzer: more input params, more tide input params to data-set.

Reply:
yes, I've used Fuzz, because the other files here have used it. I should probably rewrite it from Fuzz to quick.Check, because Fuzz doesn't support custom generators, and generation here is complicated.